### PR TITLE
Give `ValidationContext.MemberName` a distinct value from `DisplayName`

### PR DESCRIPTION
**Pull request by dougbu**
- #120

No longer true:
~~`ValidationAttribute` classes that (incorrectly) expected `MemberName` to always be set or that
instantiate `ValidationResult`s using the `DisplayName` may be negatively impacted. For exmaple,
`MemberName` will be `null` when binding parameters and items in a collection.~~

### CODE REVIEW

File: src/System.Web.Http/Validation/Validators/DataAnnotationsModelValidator.cs [hunk 1]
```diff
@@ -33,11 +33,11 @@ public override IEnumerable<ModelValidationResult> Validate(ModelMetadata metada
             // Per the WCF RIA Services team, instance can never be null (if you have
             // no parent, you pass yourself for the "instance" parameter).
             
-            string memberName = metadata.GetDisplayName();
+            var memberName = metadata.PropertyName;
```
Comment (dougbu) [inline src/System.Web.Http/Validation/Validators/DataAnnotationsModelValidator.cs]: For a while, aspnet/MVC's version of this file used `metadata.PropertyName ?? metadata.ModelType.Name`. But, though that's more like what `GetDisplayName()` did, it's also incorrect. We fixed it to look like the above in aspnet/Mvc@aecfe778ee.

Comment (dougbu) [reply src/System.Web.Http/Validation/Validators/DataAnnotationsModelValidator.cs]: I'm on the fence about this particular expression. In favour of a change: We recognized a problem and fixed that problem in ASP.NET Core MVC. Against: MVC 5.2.4 still uses `metadata.PropertyName ?? metadata.ModelType.Name` i.e. the consistency across ASP.NET argument. And, going this way makes it a larger change i.e. `MemberName` will now ignore both `[DisplayName("some name")]` and the type name for non-properties.

**Comment by dougbu** [comment]
:up::date: with more tests and less trailing whitespace

File: src/System.Web.Http/Validation/Validators/DataAnnotationsModelValidator.cs [hunk 2]
```diff
@@ -32,23 +32,23 @@ public override IEnumerable<ModelValidationResult> Validate(ModelMetadata metada
         {
             // Per the WCF RIA Services team, instance can never be null (if you have
             // no parent, you pass yourself for the "instance" parameter).
-            
-            string memberName = metadata.GetDisplayName();
+
+            var memberName = metadata.PropertyName;
```
Comment (dougbu) [inline src/System.Web.Http/Validation/Validators/DataAnnotationsModelValidator.cs]: For a while, aspnet/MVC's version of this file used `metadata.PropertyName ?? metadata.ModelType.Name`. But, though that's more like what `GetDisplayName()` did, it's also incorrect. We fixed it to look like the above in aspnet/Mvc@aecfe778ee.

I'm on the fence about this particular expression. In favour of a change: We recognized a problem and fixed that problem in ASP.NET Core MVC. Against: MVC 5.2.4 still uses `metadata.PropertyName ?? metadata.ModelType.Name` i.e. the consistency across ASP.NET argument. And, going this way makes it a larger change i.e. `MemberName` will now ignore both `[DisplayName("some name")]` and the type name for non-properties.

Comment (pranavkm) [reply src/System.Web.Http/Validation/Validators/DataAnnotationsModelValidator.cs]: Out of curiosity, when is `metadata.PropertyName` null? When we're enumerating over collection types?

Comment (dougbu) [reply src/System.Web.Http/Validation/Validators/DataAnnotationsModelValidator.cs]: That and when we're validating a parameter. IOW whenever we're not dealing w/ a property.

File: src/System.Web.Http/Validation/Validators/DataAnnotationsModelValidator.cs [hunk 3]
```diff
@@ -21,6 +29,13 @@ public DataAnnotationsModelValidator(IEnumerable<ModelValidatorProvider> validat
             Attribute = attribute;
         }
 
+        // Internal for testing
+        internal static bool UseLegacyValidationMemberName
```
Comment (dougbu) [inline src/System.Web.Http/Validation/Validators/DataAnnotationsModelValidator.cs]: `static` only because an instance field for testing seemed wasteful.

I doubt it's necessary but could make this property `public` for Microsoft.AspNet.WebApi.Owin users, whose applications are less likely to involve `System.Configuration`. Could even move this (still `static`) into `HttpConfiguration` for high visibility. Thoughts?

**Comment by dougbu** [comment]
:up::date:
/cc @Eilon since you wanted to have a gander.

File: src/System.Web.Http/Validation/Validators/DataAnnotationsModelValidator.cs (hunk 3 above)
Comment (Eilon) [reply src/System.Web.Http/Validation/Validators/DataAnnotationsModelValidator.cs]: Naah it's fine the way it is.

**Comment by dougbu** [comment]
9829dd8225
